### PR TITLE
:new: :mortar_board: Topic11 LinkedQueue implementation with LinkedQueueTest

### PR DESCRIPTION
**Pull request by jameshughes89**
### What
Add `LinkedQueue` and `LinkedQueueTest` to the repo. 

### Where
src/main/java/LinkedQueue.java
src/test/java/LinkedQueueTest.java

### Testing
Ran the tests locally.

### CODE REVIEW

File: src/test/java/LinkedQueueTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test
+    @DisplayName("Enqueuing two items results in second not being the front.")
+    void enqueuingSecondItemIsNotFront() {
+        Queue<Integer> queue = new LinkedQueue<>();
+        queue.enqueue(99);
+        queue.enqueue(101);
+        assertNotEquals(101, queue.first());
+    }
```
Comment (jameshughes89) [inline src/test/java/LinkedQueueTest.java:57]: Let me know if you think this not needed. The FIFO check would do this for us.

File: src/main/java/LinkedQueue.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        Node<T> toEnqueue = new Node<>(element);
+
+        // If nothing is in the queue, then the front is null
+        // and the enqueued element will become the front,
+        // otherwise, the enqueued element is added after the
+        // current rear.
+        if (isEmpty()) {
+            front = toEnqueue;
+        }
+        else {
```
Comment (twentylemon) [inline src/main/java/LinkedQueue.java:26]: Just fyi (and #56), I've only seen this `if/else` style from "non professional" code, for lack of a better word. You're way more likely to see the `else` inline with the `}`.

```java
if (condition) {
} else {
}
```

File: src/main/java/LinkedQueue.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        // enqueued element.
+        rear = toEnqueue;
+        size++;
+    }
+
+    @Override
+    public T dequeue() {
+        if (isEmpty()) {
+            throw new NoSuchElementException("Dequeueing from an empty queue.");
+        }
```
Comment (twentylemon) [inline src/main/java/LinkedQueue.java:39]: The exception message here is kind of useless. "No such element" already tells you the queue was empty, and the stack trace will tell you that you were in `dequeue`. It doesn't hurt, but it doesn't help.

File: src/main/java/LinkedQueue.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        T returnElement = front.getData();
+        front = front.getNext();
+        size--;
+        // Although front will update properly regardless, rear
+        // will need to be set to null if we dequeue the last
+        // element.
+        if (isEmpty()) {
+            rear = null;
+        }
```
Comment (twentylemon) [inline src/main/java/LinkedQueue.java:48]: This isn't wholly true. Don't get me wrong, you _should_ null out the `rear` for garbage collection, but adding an element will do `rear = toEnqueue`, cleaning up the internal state.

File: src/test/java/LinkedQueueTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    void enqueuingUpdatesSize() {
+        Queue<Integer> queue = new LinkedQueue<>();
+        queue.enqueue(99);
+        queue.enqueue(101);
+        assertEquals(2, queue.size());
+    }
+
+    @Test
+    @DisplayName("Enqueuing one item results in it being the front.")
+    void enqueuingOneItemIsFront() {
```
Comment (twentylemon) [inline src/test/java/LinkedQueueTest.java:44]: `front` -> `first`

File: src/test/java/LinkedQueueTest.java (hunk 1 above)
Comment (twentylemon) [reply src/test/java/LinkedQueueTest.java:57]: You may have answered your own question already. What's actually being tested here? If this fails, does the FIFO always tests fail? If the FIFO test fails, will this always fail?

File: src/main/java/LinkedQueue.java (hunk 3 above)
Comment (jameshughes89) [reply src/main/java/LinkedQueue.java:39]: Message is just to show them that it's a thing.

File: src/main/java/LinkedQueue.java (hunk 4 above)
Comment (jameshughes89) [reply src/main/java/LinkedQueue.java:48]: I added a note about this in #112 mentioning not doing this and messing up a count. 

To include your point however, I will also add a mention of this being good for garbage collection to #112 and change this comment to _should_ from _will need to_.

**Review by twentylemon — APPROVED** [review]

**Pull request merged**